src/ConnectedMode.UnitTests/BranchMatcherTests.cs

```diff
@@ -0,0 +1,211 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using SonarLint.VisualStudio.ConnectedMode.UnitTests.LibGit2SharpWrappers;
+using SonarQube.Client;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.UnitTests
+{
+    [TestClass]
+    public class BranchMatcherTests
+    {
+        [TestMethod]
+        public async Task GetMatchedBranch_ChooseBranchWithSameName()
+        {
+            var service = CreateService("master", "branch");
+
+            var masterCommit = new CommitWrapper(1);
+            var branchCommit = new CommitWrapper(2);
+
+            var masterBranch = CreateBranch("master", masterCommit);
+            var headBranch = CreateBranch("branch", branchCommit, masterCommit);
+
+            var repo = CreateRepo(headBranch, masterBranch);
+
+            var testSubject = new BranchMatcher(service, repo);
+
+            var result = await testSubject.GetMatchedBranch("projectKey");
+
+            result.Should().Be("branch");
+        }
+
+        [TestMethod]
+        public async Task GetMatchedBranch_NoBranchWithSameName_ChooseClosestMatch()
+        {
+            var service = CreateService("master", "dev");
+
+            var masterCommit = new CommitWrapper(1);
+            var devBranchCommit = new CommitWrapper(2);
+            var headBranchCommit = new CommitWrapper(3);
+
+            var masterBranch = CreateBranch("master", masterCommit);
+            var devBranch = CreateBranch("dev", devBranchCommit, masterCommit);
+            var headBranch = CreateBranch("branch", headBranchCommit, devBranchCommit, masterCommit);
+
+            var repo = CreateRepo(headBranch, masterBranch, devBranch);
+
+            var testSubject = new BranchMatcher(service, repo);
+
+            var result = await testSubject.GetMatchedBranch("projectKey");
+
+            result.Should().Be("dev");
+        }
+
+        /*
+         * Master
+         * |->Commit1
+         * |
+         * |->Commit2
+         * |
+         * |->Commit3
+         * |    |
+         * |    |---->Branch1
+         * |    |       |
+         * |    |       |->Commit4
+         * |    |       |
+         * |    |       |->Commit5
+         * |    |       |
+         * |    |
+         * |    |---->Branch2 (closest branch)
+         * |    |       |
+         * |    |       |->Commit6
+         * |    |       |
+         * |    |
+         * |    |---->Branch3 (Head)
```

Comment:
Could you make the test slightly more robust by including a closest commit match that does not have a matching branch name?
i.e. a local branch that is closest, but that does not match a branch on the server -> should be ignored.

Does that make sense as an additional test case?

Reply:
Then it would match itself. But it makes sense to create such a test to be explicit.  (maybe not in this complicated format)

Reply:
Why would it match itself? It's a local branch that doesn't have a server match so it should be ignored, regardless of how close it is.